src/webcore/promise.rs

```diff
@@ -0,0 +1,260 @@
+use std;
+use std::error::Error as _Error;
+use std::marker::PhantomData;
+use webcore::once::Once;
+use webcore::value::{Value, Reference, ConversionError};
+use webcore::try_from::{TryInto, TryFrom};
+use web::error::Error;
+use futures::{Future, Poll, Async};
+use futures::unsync::oneshot::{Receiver, channel};
+use webcore::promise_executor::spawn;
+
+
+/// A `Promise` object represents the eventual completion (or failure) of an asynchronous operation, and its resulting value.
+///
+/// In most situations you shouldn't use this, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
+// https://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects
+#[derive(Clone, Debug, ReferenceType)]
+#[reference(instance_of = "Promise")]
+pub struct Promise( Reference );
+
+impl Promise {
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise-resolve-functions
+    fn is_promise_like( input: &Value ) -> bool {
+        (js! {
+            var input = @{input};
+            // This emulates the `Type(input) is Object` and `IsCallable(input.then)` ECMAScript abstract operations.
+            return Object( input ) === input &&
+                   typeof input.then === "function";
+        }).try_into().unwrap()
+    }
+
+    /// This function should rarely be needed, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+    ///
+    /// This function is needed if you have a JavaScript value which is a Promise-like object
+    /// (it has a `then` method) but it isn't a true `Promise`.
+    ///
+    /// That situation is rare, but it can happen if you are using a Promise library such as jQuery or
+    /// Bluebird.
+    ///
+    /// In that situation you can use `Promise::convert(value)` to convert it into a true `Promise`.
+    ///
+    /// If the `input` isn't a Promise-like object then it returns `None`.
+    ///
+    /// # Examples
+    ///
+    /// Convert a Promise-like object to a `Promise`:
+    ///
+    /// ```rust
+    /// // jQuery Promise
+    /// Promise::convert(js!( return $.get("test.php"); ))
+    ///
+    /// // Bluebird Promise
+    /// Promise::convert(js!( return bluebird_promise.timeout(1000); ))
+    /// ```
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve)
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise.resolve
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise-resolve-functions
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promiseresolvethenablejob
+    pub fn convert( input: Value ) -> Option< Self > {
```

Comment:
But if it accepts `Reference` then you would need to do something like this:

```rust
Promise::convert(js!( return ... ).try_into().unwrap())
```

...which is annoying, verbose, and slower. And it doesn't really help, because even with a `Reference` it *still* might return `None`, so you don't gain any real static type safety.

Also, I've noticed that sometimes objects get turned into `Object`, not `Reference`, and the `Promise::convert` function works with *any* object which happens to have a `then` method (even Arrays!). So I think using `Value` is correct.

Reply:
> Also, I've noticed that sometimes objects get turned into Object (not Reference)

This is actually a very good point, thank you! Hmm...

The original idea was that `Object` objects represent a plain, dumb, JavaScript object for holding data, but unfortunately a lot of "fat" objects (e.g. like Bluebird promises) in the wild seem to be indistinguishable from those which are just simple `{key: value}` containers for data. I'm thinking that perhaps we should stop treating `Object`s and `Array`s in a special way, make them a normal `#[derive(ReferenceType)]` types and only have a `Reference` variant in `Value`.

I'm starting to really like this idea; then the problem of promises (and other kinds of objects) being converted to an `Object` will disappear.

> And it doesn't really help, because even with a Reference it still might return None, so you don't gain any real static type safety.

(Assuming we'll remove `Value::Object` variant.)

1. Yes, but, conversely if someone already has a `Reference` they have to convert it to a `Value` now to pass into this function.
2. The types are there not only for guaranteeing type safety, but also to serve as documentation.
3. Casting a `Value` to a `Reference` is relatively cheap as it's just a `match`.
4. You could also convert it this way: `js!( ... ).as_reference().and_then( Promise::convert ).unwrap()`, which I don't think is that bad?

Reply:
If we remove the `Object` and `Array` variants, then yes I agree it would make more sense to make it `Reference`. But it's still a lot more verbose than using `Value`, so I'm not sure.

I'll leave it as `Value` for now, with a TODO note to change it later (if we decide to remove `Object` and `Array` variants).